File: Bookcase web/src/test/java/cz/vhromada/book/web/common/CategoryUtils.java
package cz.vhromada.book.web.common;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.SoftAssertions.assertSoftly;

import java.util.List;

import cz.vhromada.book.entity.Category;
import cz.vhromada.book.web.fo.CategoryFO;

/**
 * A class represents utility class for categories.
 *
 * @author Vladimir Hromada
 */
public final class CategoryUtils {

    /**
     * Creates a new instance of CategoryUtils.
     */
    private CategoryUtils() {
    }

    /**
     * Returns FO for category.
     *
     * @return FO for category
     */
    public static CategoryFO getCategoryFO() {
        final CategoryFO category = new CategoryFO();
        category.setId(1);
        category.setName("Name");
        category.setPosition(0);

        return category;
    }

    /**
     * Returns category.
     *
     * @return category
     */
    public static Category getCategory() {
        final Category category = new Category();
        category.setId(1);
        category.setName("Name");
        category.setPosition(0);

        return category;
    }

    /**
     * Asserts category deep equals.
     *
     * @param expected expected FO for category
     * @param actual   actual category
     */
    public static void assertCategoryDeepEquals(final CategoryFO expected, final Category actual) {
        assertSoftly(softly -> {
            softly.assertThat(expected).isNotNull();
            softly.assertThat(actual).isNotNull();
        });
        assertSoftly(softly -> {
            softly.assertThat(actual.getId()).isEqualTo(expected.getId());
            softly.assertThat(actual.getName()).isEqualTo(expected.getName());
            softly.assertThat(actual.getPosition()).isEqualTo(expected.getPosition());
        });
    }

    /**
     * Asserts categories deep equals.
     *
     * @param expected expected list of categories
     * @param actual   actual list of category
     */
    public static void assertCategoriesDeepEquals(final List<Integer> expected, final List<Category> actual) {
        assertSoftly(softly -> {
            softly.assertThat(expected).isNotNull();
            softly.assertThat(actual).isNotNull();
        });
        assertThat(actual.size()).isEqualTo(expected.size());
        for (int i = 0; i < expected.size(); i++) {
            assertCategoryDeepEquals(expected.get(i), actual.get(i));
        }
    }

    /**
     * Asserts category deep equals.
     *
     * @param expected expected category
     * @param actual   actual category
     */
    public static void assertCategoryDeepEquals(final Integer expected, final Category actual) {
        assertSoftly(softly -> {
            softly.assertThat(expected).isNotNull();
            softly.assertThat(actual).isNotNull();
        });
        assertSoftly(softly -> {
            softly.assertThat(actual.getId()).isEqualTo(expected);
            softly.assertThat(actual.getName()).isNull();
            softly.assertThat(actual.getPosition()).isNull();
        });
    }

    /**
     * Asserts categories deep equals.
     *
     * @param expected expected list of category
     * @param actual   actual list of categories
     */
    public static void assertCategoryListDeepEquals(final List<Category> expected, final List<Integer> actual) {
        assertSoftly(softly -> {
            softly.assertThat(expected).isNotNull();
            softly.assertThat(actual).isNotNull();
        });
        assertThat(actual.size()).isEqualTo(expected.size());
        for (int i = 0; i < expected.size(); i++) {
            assertCategoryDeepEquals(expected.get(i), actual.get(i));
        }
    }

    /**
     * Asserts category deep equals.
     *
     * @param expected expected category
     * @param actual   actual category
     */
    public static void assertCategoryDeepEquals(final Category expected, final Integer actual) {
        assertSoftly(softly -> {
            softly.assertThat(expected).isNotNull();
            softly.assertThat(actual).isNotNull();
        });
        assertSoftly(softly -> softly.assertThat(actual).isEqualTo(expected.getId()));
    }

}

File: bookcase-web/src/test/kotlin/cz/vhromada/bookcase/web/common/CategoryUtils.kt
package cz.vhromada.bookcase.web.common

import cz.vhromada.bookcase.entity.Category
import cz.vhromada.bookcase.web.fo.CategoryFO
import org.assertj.core.api.Assertions
import org.assertj.core.api.SoftAssertions.assertSoftly

/**
 * A class represents utility class for categories.
 *
 * @author Vladimir Hromada
 */
object CategoryUtils {

    /**
     * Returns FO for category.
     *
     * @return FO for category
     */
    fun getCategoryFO(): CategoryFO {
        return CategoryFO(
                id = 1,
                name = "Name",
                position = 0)
    }

    /**
     * Returns category.
     *
     * @return category
     */
    fun getCategory(): Category {
        return Category(
                id = 1,
                name = "Name",
                position = 0)
    }

    /**
     * Asserts category deep equals.
     *
     * @param expected expected FO for category
     * @param actual   actual category
     */
    fun assertCategoryDeepEquals(expected: CategoryFO?, actual: Category?) {
        assertSoftly {
            it.assertThat(expected).isNotNull
            it.assertThat(actual).isNotNull
        }
        assertSoftly {
            it.assertThat(actual!!.id).isEqualTo(expected!!.id)
            it.assertThat(actual.name).isEqualTo(expected.name)
            it.assertThat(actual.position).isEqualTo(expected.position)
        }
    }

    /**
     * Asserts categories deep equals.
     *
     * @param expected expected list of categories
     * @param actual   actual list of categories
     */
    fun assertCategoriesDeepEquals(expected: List<Category?>?, actual: List<Int?>?) {
        assertSoftly {
            it.assertThat(expected).isNotNull
            it.assertThat(actual).isNotNull
        }
        Assertions.assertThat(actual!!.size).isEqualTo(expected!!.size)
        for (i in expected.indices) {
            assertCategoryDeepEquals(expected[i], actual[i])
        }
    }

    /**
     * Asserts category deep equals.
     *
     * @param expected expected category
     * @param actual   actual category
     */
    private fun assertCategoryDeepEquals(expected: Category?, actual: Int?) {
        assertSoftly {
            it.assertThat(expected).isNotNull
            it.assertThat(actual).isNotNull
        }
        assertSoftly { it.assertThat(actual).isEqualTo(expected!!.id) }
    }

}
